### ISSUE TITLE
Adding PaymentRequest API support

### ISSUE THREAD

**Issue by madmath** [woocommerce/woocommerce#12522]
Would it be possible to support the PaymentRequest API for faster checkouts?

Most browsers have committed to implement it, with (sometimes preliminary) support having landed in Edge, Chrome, Opera, Firefox. 

Web Payments API W3C spec: https://w3c.github.io/browser-payment-api/

Browser support:
Edge
https://twitter.com/RReverser/status/804451484758749184

Chrome (shipping to all users)
https://developers.google.com/web/updates/2016/07/payment-request

Firefox
https://github.com/mozilla-payments/

**Comment by mikejolley** [comment]
PaymentRequest API does not do anything on it's own - it requires a gateway to facilitate the payment. We'll consider adding to core when support widens, but for now we're committed only to shipping this with certain extensions that support the API.

Stripe has support for Chrome PaymentRequest API implementation right now https://github.com/woocommerce/woocommerce-gateway-stripe

**Comment by jonathanbardo** [comment]
@mikejolley I feel this should be left open for wider consideration. The payment request API does simplify the checkout process a lot by exposing a native UI and should be supported in core in my opinion. I think the payment gateway plugin should explicitly tell core if it supports that feature or not.

**Comment by madmath** [comment]
Thanks Mike for the link, interesting to see Stripe support this!

Given that Payment Request API is agnostic to the payment processor and amongst various things supports "raw" credit cards the same way a credit card <form> would, it probably makes sense to have the core engine support it. But I don't know enough about your code structure; feel free to dismiss this :)

**Comment by mikejolley** [comment]
@jonathanbardo You're assuming we're not considering it :) Rolling into core is the eventual plan, but we're building it into some gateway extensions first. i.e. to test the feature before rolling it into core. You'll notice Stripe extension also supports Apple Pay, and we have begun looking into Edge support too.

Since it's still in draft, it doesn't make sense to add that to core until things settle. Experimental features should go into feature plugins, or the extensions themselves.

@madmath yes and no. Core doesn't have a credit card processor built in, so it would literally be unused code that we cannot iterate on as fast as we can in an extension. 

This API also has implications on other aspects of WC such as bypassing our checkout flow (meaning some checkout-based extensions never run). Plus we need to work out how to handle multiple gateways at once when say one does not support it. And to decide whether to support triggers from regular checkout rather than just cart page.

**tl;dr** Once the spec is more mature (it's a draft!) we'll consider moving it out of Stripe/extensions and into core. Until then, you can still make use of the improvements with an extension :)

If you have any feedback on the Stripe extension's implementation, please raise in that repo.

**Comment by madmath** [comment]
Thanks for the detailed reply, Mike! 

One note: Given that browser vendors are collaborating on the spec, the Edge implementation will be no different than Chrome's or Opera's, for example! Yay standards :)

**Comment by mikejolley** [comment]
@madmath Ok, maybe I'm just a bit paranoid from past experiences :D
